keras_cv/models/segmentation/deeplab_v3/deeplab_v3.py

```diff
@@ -0,0 +1,201 @@
+                "backbone.input_shape={backbone.input_shape[1:]}"
+            )
+
+        height = input_shape[0]
+        width = input_shape[1]
+
+        feature_map = backbone(inputs)
+        if spatial_pyramid_pooling is None:
+            spatial_pyramid_pooling = SpatialPyramidPooling(
+                dilation_rates=[6, 12, 18]
+            )
+
+        outputs = spatial_pyramid_pooling(feature_map)
+        outputs = keras.layers.UpSampling2D(
+            size=(
+                height // feature_map.shape[1],
+                width // feature_map.shape[2],
+            ),
+            interpolation="bilinear",
+        )(outputs)
+
+        if segmentation_head is None:
+            segmentation_head = SegmentationHead(
+                num_classes=num_classes,
+                name="segmentation_head",
+                convolutions=1,
+                dropout=0.2,
+                kernel_size=1,
+                activation=segmentation_head_activation,
+            )
+
+        # Segmentation head expects a multiple-level output dictionary
+        outputs = segmentation_head({1: outputs})
+
+        super().__init__(inputs=inputs, outputs=outputs, **kwargs)
+
+        self.num_classes = num_classes
+        self.backbone = backbone
+        self.spatial_pyramid_pooling = spatial_pyramid_pooling
+        self.segmentation_head = segmentation_head
+        self.segmentation_head_activation = segmentation_head_activation
+
+    def get_config(self):
+        return {
+            "num_classes": self.num_classes,
+            "backbone": self.backbone,
+            "spatial_pyramid_pooling": self.spatial_pyramid_pooling,
+            "segmentation_head": self.segmentation_head,
+            "segmentation_head_activation": self.segmentation_head_activation,
+        }
+
+    def build(self, input_shape):
+        height = input_shape[1]
+        width = input_shape[2]
+        feature_map_shape = self.backbone.compute_output_shape(input_shape)
+        self.up_layer = keras.layers.UpSampling2D(
+            size=(
+                height // feature_map_shape[1],
+                width // feature_map_shape[2],
+            ),
+            interpolation="bilinear",
+        )
+
+    def compile(self, weight_decay=0.0001, **kwargs):
+        """compiles the DeepLabV3 model.
+
+        Args:
+            weight_decay: Optional float, factor of weight decay applied during
+                training, defaults to `0.0001`.
+        """
+        self.weight_decay = weight_decay
+        super().compile(**kwargs)
+
+    def train_step(self, data):
+        images, y_true, sample_weight = keras.utils.unpack_x_y_sample_weight(
+            data
+        )
+        with tf.GradientTape() as tape:
+            y_pred = self(images, training=True)
+            total_loss = self.compute_loss(
+                images, y_true, y_pred, sample_weight
+            )
+            reg_losses = []
+            if self.weight_decay:
```

Comment:
Should it be used here? Or, via optimizer!

Reply:
+1 -- we should drop this from the model. This was originally written at a time where weight decay wasn't available in all optimizers

Reply:
(I think this should mean we shouldn't need a custom train_Step at all, or a custom compile)

Reply:
I think if we want to always ensure L2 regularization during training only then it makes sense. However, I think doing it via the optimizer makes more sense.